### PR TITLE
DAOS-6373 test: refactor datamover test infrastructure

### PR DESCRIPTION
**Pull request by daltonbohning**
This is a massive overhaul of the datamover functional test framework. It might be easier to "view file" instead of looking at the diffs. I'll need to run this a couple more times, but I would appreciate any feedback now.

- Renamed DataMoverCommand to DcpCommand
- Overhaul of datamover test utils and test base
  - Log the number of times run_datamover is called
  - Switch between multiple utilities
  - run_datamover now allows setting params
  - replaced set_src and set_dst with set_datamover_params
  - added mdtest support to datamover test base
- Updated tests to match test base changes
- Updated test documentation to make it clear which test
cases are for POSIX copies.
- Updated some timeouts
- Updated avocado tag formats to be consistent
- Updated copy_basics to run more variations
- Removed the test copy_auto_create_dest
- Updated container_copy.py to use DataMoverTestBase
- Corrected issue in container_copy.py where the destination
was incorrect.
- Refactored container_copy.py into copy_large_ior and copy_large_mdtest
  - This was necessary because there was a !mux for just mdtest
  - Previously, the ior test was running twice because of !mux

mdtest_test_base.py
- Added optional pool and display_space arguments to run_mdtest.

Signed-off-by: Dalton Bohning <daltonx.bohning@intel.com>

### CODE REVIEW

**Review by shimizukko — APPROVED** [review]
LGTM
What is dcp? Is it the same as datamover?

**Comment by daltonbohning** [comment]
> LGTM
> What is dcp? Is it the same as datamover?

@shimizukko "datamover" is more of an umbrella term that refers to multiple utilities. There are the utilities that do and will exist:
dcp - "distributed" cp - From mpifileutils. "cp" using MPI
dsync - "distributed" - From mpifileutils.  "sync" using MPI
daos filesystem copy - From "daos" command - Copy between POSIX containers and POSIX filesystems
daos container copy - From "daos" command - Copy between any two DAOS containers at the obj level

The reason for the rename from "datamover" to "dcp" is to be more explicit when referring to the tool to use. And each tool has slightly different options, so a single "datamover" command class can't easily encompass all utilities. Ultimately, each of these utilities will have a "command" class to execute them.

**Comment by shimizukko** [comment]
> > LGTM
> > What is dcp? Is it the same as datamover?
> 
> @shimizukko "datamover" is more of an umbrella term that refers to multiple utilities. There are the utilities that do and will exist:
> dcp - "distributed" cp - From mpifileutils. "cp" using MPI
> dsync - "distributed" - From mpifileutils. "sync" using MPI
> daos filesystem copy - From "daos" command - Copy between POSIX containers and POSIX filesystems
> daos container copy - From "daos" command - Copy between any two DAOS containers at the obj level
> 
> The reason for the rename from "datamover" to "dcp" is to be more explicit when referring to the tool to use. And each tool has slightly different options, so a single "datamover" command class can't easily encompass all utilities. Ultimately, each of these utilities will have a "command" class to execute them.
> 
> If you're curious, here is the in-progress documentation:
> https://wiki.hpdd.intel.com/display/DAOS/Data+Mover+Documentation

I see. Thanks for the info. There are quite new to me.

**Comment by daltonbohning** [comment]
> I see. Thanks for the info. There are quite new to me

You're welcome! They are new to most people, so no worries there :)

**Comment by daltonbohning** [comment]
@saurabhtandan @shimizukko I've ran this a few times now, and the changes seem to be stable.

**Review by saurabhtandan — COMMENTED** [review]
Code wise looks good to me but I was not able to find test runs for CopyLargeIor, CopyLargeMdtest and CopyNegative. I will hold my review to see how the current PR run goes.

**Comment by daltonbohning** [comment]
> Code wise looks good to me but I was not able to find test runs for CopyLargeIor, CopyLargeMdtest and CopyNegative. I will hold my review to see how the current PR run goes.

Interesting...
This previous run has those results:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4236/3/tests
But newer runs do not.
It seems the "Features" tag had some changes:
https://github.com/daos-stack/pipeline-lib/blob/f852f4e7efa77b6815bf331148db71bf3d500fa5/vars/parseStageInfo.groovy#L155-L157

I will re-run this with the tests manually put in.
Thanks for pointing that out!

File: src/tests/ftest/util/data_mover_test_base.py
Comment (shimizukko) [inline src/tests/ftest/util/data_mover_test_base.py:354]:
```suggestion
                Alternatively, this can be the pool uuid.
```

File: src/tests/ftest/util/data_mover_test_base.py
Comment (shimizukko) [inline src/tests/ftest/util/data_mover_test_base.py:362]:
```suggestion
                Alternatively, this can be the pool uuid.
```

**Review by shimizukko — APPROVED** [review]
LGTM

**Comment by daltonbohning** [comment]
NLT Stage is yellow - should be fixed by https://github.com/daos-stack/daos/pull/4304

DaosCoreTest (test_T) is failing. This PR does not affect that test, and that test is failing elsewhere. I will create a ticket if needed, once this run completes.
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4236/9/tests

File: src/tests/ftest/util/data_mover_test_base.py
Comment (daltonbohning) [reply src/tests/ftest/util/data_mover_test_base.py:354]: Oh, that's embarrassing...

**Comment by daltonbohning** [comment]
All datamover tests are passing:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4236/10/tests

Known Failures:
NLT stage is yellow - unrelated
DaosCoreTest (test_T) - unrelated, these tests do not use datamover, and this test is failing elsewhere

[DAOS-6484](https://jira.hpdd.intel.com/browse/DAOS-6484) documents a failure for the copy_basics test on Leap15, but that issue has yet to be seen on this PR.

**Comment by daltonbohning** [comment]
All datamover tests are passing:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4236/10/tests

[DAOS-6484](https://jira.hpdd.intel.com/browse/DAOS-6484) disabled the copy_types test, so I also re-disabled copy_subsets, since it could fail for the same reason.

All enabled datamover tests are passing:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4236/11/pipeline

**Comment by daltonbohning** [comment]
@shimizukko @saurabhtandan This is ready for review now, with all enabled datamover tests passing.

**Review by shimizukko — DISMISSED** [review]
LGTM

**Comment by daltonbohning** [comment]
This PR conflicts with Ken's svc PR: https://github.com/daos-stack/daos/pull/3935 for DAOS-4433
It would be best to wait until that PR is merged to master, and then resolve conflicts here.

**Comment by daltonbohning** [comment]
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-4236/14/tests
Latest run passes all enabled datamover functional tests.
There are a lot of cart failures which are unrelated, and I believe are known failures (but there are a lot to track down).

**Review by shimizukko — DISMISSED** [review]
LGTM

**Review by saurabhtandan — APPROVED** [review]
LGTM

File: src/tests/ftest/datamover/copy_basics.yaml
Comment (daltonbohning) [reply src/tests/ftest/datamover/copy_basics.yaml:33]: Ah, yes. I was originally pushing some fs copy tests as well, but backed those out.
This will be corrected in my next PR.

File: src/tests/ftest/datamover/copy_large_mdtest.yaml
Comment (daltonbohning) [reply src/tests/ftest/datamover/copy_large_mdtest.yaml:56]: This will be corrected in my next PR.

**Comment by daltonbohning** [comment]
Had to resolve yet another conflict in master. Will need to wait for results.

**Review by shimizukko — APPROVED** [review]
LGTM

**Review by saurabhtandan — APPROVED** [review]
LGTM

**Pull request merged**
